plugin/handler_socket/handlersocket/database.cpp

```diff
@@ -304,6 +304,7 @@ dbcontext::init_thread(const void *stack_bottom, volatile int& shutdown_flag)
       thd->db = 0;
       thd->db = my_strdup("handlersocket", MYF(0));
     }
+    thd->variables.option_bits |= OPTION_TABLE_LOCK;
```

Comment:
Since rev. 9bc5cec0f115d7d0c277a49b91b96109560280f4 we allocate locks for regular statements on THD::mem_root. In contrast locks acquired by LOCK TABLES must survive along multiple statements and thus are allocated on heap (when OPTION_TABLE_LOCK is in effect).

Handlersocket doesn't seem to reset THD::mem_root, which causes this memory leak. Setting OPTION_TABLE_LOCK is an options, but it looks fragile and not performance wise: resetting THD::mem_root should be the right way to fix this.